[C23] Implement WG14 N3037

AaronBallman opened the pull request:
This changes the type compatibility rules so that it is permitted to redefine tag types within the same TU so long as they are equivalent definitions.

It is intentionally not being exposed as an extension in older C language modes. GCC does not do so and the feature doesn't seem compelling enough to warrant it.

AaronBallman committed:
[C23] Implement WG14 N3037

This changes the type compatibility rules so that it is permitted to
redefine tag types within the same TU so long as they are equivalent
definitions.

AaronBallman committed:
Start using structural equivalence checking.

AaronBallman committed:
Handle enumerations

AaronBallman committed:
Handle attributes, add a ton of tests

AaronBallman committed:
Update website

AaronBallman committed:
Fix failing tests

AaronBallman committed:
Add additional test cases

and simplify the implementation somewhat.

AaronBallman committed:
Add codegen test

AaronBallman committed:
Add an enum test with a fixed underlying type

AaronBallman committed:
Properly diagnose bit-fields

AaronBallman committed:
Improve diagnostic consistency

AaronBallman committed:
Fix failing tests

AaronBallman committed:
Add another test, crank up diagnostic levels

AaronBallman committed:
Even more test cases

Sirraide commented on clang/test/C/C23/n3037.c:
```diff
@@ -0,0 +1,456 @@
+struct hidden_struct { // This is fine because the previous declaration is not visible.
+  int y;
+  int z;
+};
```
Some additional tests with `alignas` maybe? Though iirc we model that as an attribute so that should just work.

AaronBallman replied:
I'll add test coverage

AaronBallman commented on clang/test/C/C23/n3037.c:
```diff
@@ -0,0 +1,456 @@
+
+struct food { int (*p)[3]; }; // c23-note {{field 'p' has type 'int (*)[3]' here}} \
+                                 c17-note {{previous definition is here}}
+struct food { int (*p)[]; };  // c23-error {{type 'struct food' has incompatible definitions}} \
```
They're not, I'll add this as a test case.

AaronBallman committed:
Add additional test coverage

AaronBallman committed:
Restore code that was still used, just not in an obvious way

AaronBallman commented on clang/lib/AST/ASTStructuralEquivalence.cpp:
```diff
@@ -450,6 +453,116 @@ class StmtComparer {
+  llvm::sort(Attrs1, Sorter);
+  llvm::sort(Attrs2, Sorter);
+
+  auto A2 = Attrs2.begin(), A2End = Attrs2.end();
+  const auto *DiagnoseDecl = cast<TypeDecl>(PrimaryDecl ? PrimaryDecl : D2);
+  for (auto A1 = Attrs1.begin(), A1End = Attrs1.end(); A1 != A1End;
+       ++A1, ++A2) {
+    if (A2 == A2End) {
```
I thought about using `zip` but didn't really like the direction it was going. We need to know three things: 1) attribute on lhs, no attribute on rhs, 2) no attribute on lhs, attribute on rhs, 3) different attribute on lhs than on rhs. `zip` makes 1 and 2 more awkward in some ways; so I went with the same code pattern as the original code had.

WDYT?

Sirraide replied:
Ah, yeah, makes sense

tbaederr commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -11432,6 +11433,22 @@ static QualType mergeEnumWithInteger(ASTContext &Context, const EnumType *ET,
+  // C23, on the other hand, requires the members to be "the same enough", so
+  // we use a structural equivalence check.
+  StructuralEquivalenceContext::NonEquivalentDeclSet NonEquivalentDecls;
+  StructuralEquivalenceContext Ctx(
+      getLangOpts(), *this, *this, NonEquivalentDecls,
+      StructuralEquivalenceKind::Default, false /*StrictTypeSpelling*/,
+      false /*Complain*/, true /*ErrorOnTagTypeMismatch*/);
+  return Ctx.IsEquivalent(LHS, RHS) ? LHS : QualType{};
```
```suggestion
  StructuralEquivalenceContext Ctx(
      getLangOpts(), *this, *this, NonEquivalentDecls,
      StructuralEquivalenceKind::Default,  /*StrictTypeSpelling=*/false,
       /*Complain=*/false,  /*ErrorOnTagTypeMismatch=*/true);
  return Ctx.IsEquivalent(LHS, RHS) ? LHS : QualType{};
```

AaronBallman committed:
Change formatting based on review feedback

shafik commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -11432,6 +11433,22 @@ static QualType mergeEnumWithInteger(ASTContext &Context, const EnumType *ET,
+  StructuralEquivalenceContext::NonEquivalentDeclSet NonEquivalentDecls;
+  StructuralEquivalenceContext Ctx(
+      getLangOpts(), *this, *this, NonEquivalentDecls,
+      StructuralEquivalenceKind::Default, /*StrictTypeSpelling*/ false,
+      /*Complain*/ false, /*ErrorOnTagTypeMismatch*/ true);
```
```suggestion
      StructuralEquivalenceKind::Default, /*StrictTypeSpelling=*/false,
      /*Complain=*/false, /*ErrorOnTagTypeMismatch*=/true);
```

shafik commented on clang/lib/Sema/SemaType.cpp:
```diff
@@ -9089,8 +9089,8 @@ bool Sema::hasStructuralCompatLayout(Decl *D, Decl *Suggested) {
-      StructuralEquivalenceKind::Default,
+      getLangOpts(), D->getASTContext(), Suggested->getASTContext(),
+      NonEquivalentDecls, StructuralEquivalenceKind::Default,
       false /*StrictTypeSpelling*/, true /*Complain*/,
       true /*ErrorOnTagTypeMismatch*/);
```
```suggestion
       /*StrictTypeSpelling=*/false, /*Complain=*/true,
       /*ErrorOnTagTypeMismatch=*/true);
```

shafik commented on clang/unittests/AST/StructuralEquivalenceTest.cpp:
```diff
@@ -1849,8 +1852,9 @@ TEST_F(StructuralEquivalenceCacheTest, ReturnStmtNonEq) {
-      NonEquivalentDecls, StructuralEquivalenceKind::Default, false, false);
+      get<0>(TU)->getASTContext().getLangOpts(), get<0>(TU)->getASTContext(),
+      get<1>(TU)->getASTContext(), NonEquivalentDecls,
+      StructuralEquivalenceKind::Default, false, false);
```
```suggestion
      StructuralEquivalenceKind::Default, /*StrictTypeSpelling=*/false, /*Complain=*/false);
```

shafik commented on clang/unittests/AST/StructuralEquivalenceTest.cpp:
```diff
@@ -1826,8 +1828,9 @@ TEST_F(StructuralEquivalenceCacheTest, SimpleNonEq) {
-      NonEquivalentDecls, StructuralEquivalenceKind::Default, false, false);
+      get<0>(TU)->getASTContext().getLangOpts(), get<0>(TU)->getASTContext(),
+      get<1>(TU)->getASTContext(), NonEquivalentDecls,
+      StructuralEquivalenceKind::Default, false, false);
```
```suggestion
      StructuralEquivalenceKind::Default, /*StrictTypeSpelling=*/false, /*Complain=*/false);
```

shafik commented on clang/unittests/AST/StructuralEquivalenceTest.cpp:
```diff
@@ -1868,8 +1872,9 @@ TEST_F(StructuralEquivalenceCacheTest, VarDeclNoEq) {
-      NonEquivalentDecls, StructuralEquivalenceKind::Default, false, false);
+      get<0>(TU)->getASTContext().getLangOpts(), get<0>(TU)->getASTContext(),
+      get<1>(TU)->getASTContext(), NonEquivalentDecls,
+      StructuralEquivalenceKind::Default, false, false);
```
```suggestion
      StructuralEquivalenceKind::Default, /*StrictTypeSpelling=*/false, /*Complain=*/false);
```

shafik commented on clang/unittests/AST/StructuralEquivalenceTest.cpp:
```diff
@@ -1886,8 +1891,9 @@ TEST_F(StructuralEquivalenceCacheTest, VarDeclWithDifferentStorageClassNoEq) {
-      NonEquivalentDecls, StructuralEquivalenceKind::Default, false, false);
+      get<0>(TU)->getASTContext().getLangOpts(), get<0>(TU)->getASTContext(),
+      get<1>(TU)->getASTContext(), NonEquivalentDecls,
+      StructuralEquivalenceKind::Default, false, false);
```
```suggestion
      StructuralEquivalenceKind::Default, /*StrictTypeSpelling=*/false, /*Complain=*/false);
```

shafik commented on clang/unittests/AST/StructuralEquivalenceTest.cpp:
```diff
@@ -1913,8 +1919,9 @@ TEST_F(StructuralEquivalenceCacheTest,
-      NonEquivalentDecls, StructuralEquivalenceKind::Default, false, false);
+      get<0>(TU)->getASTContext().getLangOpts(), get<0>(TU)->getASTContext(),
+      get<1>(TU)->getASTContext(), NonEquivalentDecls,
+      StructuralEquivalenceKind::Default, false, false);
```
```suggestion
      StructuralEquivalenceKind::Default, /*StrictTypeSpelling=*/false, /*Complain=*/false);
```

shafik commented on clang/unittests/AST/StructuralEquivalenceTest.cpp:
```diff
@@ -1932,8 +1939,9 @@ TEST_F(StructuralEquivalenceCacheTest, VarDeclWithInitNoEq) {
-      NonEquivalentDecls, StructuralEquivalenceKind::Default, false, false);
+      get<0>(TU)->getASTContext().getLangOpts(), get<0>(TU)->getASTContext(),
+      get<1>(TU)->getASTContext(), NonEquivalentDecls,
+      StructuralEquivalenceKind::Default, false, false);
```
```suggestion
      StructuralEquivalenceKind::Default, /*StrictTypeSpelling=*/false, /*Complain=*/false);
```

shafik commented on clang/unittests/AST/StructuralEquivalenceTest.cpp:
```diff
@@ -1964,8 +1972,9 @@ TEST_F(StructuralEquivalenceCacheTest, SpecialNonEq) {
-      NonEquivalentDecls, StructuralEquivalenceKind::Default, false, false);
+      get<0>(TU)->getASTContext().getLangOpts(), get<0>(TU)->getASTContext(),
+      get<1>(TU)->getASTContext(), NonEquivalentDecls,
+      StructuralEquivalenceKind::Default, false, false);
```
```suggestion
      StructuralEquivalenceKind::Default, /*StrictTypeSpelling=*/false, /*Complain=*/false);
```

shafik commented on clang/unittests/AST/StructuralEquivalenceTest.cpp:
```diff
@@ -2003,8 +2012,9 @@ TEST_F(StructuralEquivalenceCacheTest, Cycle) {
-      NonEquivalentDecls, StructuralEquivalenceKind::Default, false, false);
+      get<0>(TU)->getASTContext().getLangOpts(), get<0>(TU)->getASTContext(),
+      get<1>(TU)->getASTContext(), NonEquivalentDecls,
+      StructuralEquivalenceKind::Default, false, false);
```
```suggestion
      StructuralEquivalenceKind::Default, /*StrictTypeSpelling=*/false, /*Complain=*/false);
```

shafik commented on clang/unittests/AST/StructuralEquivalenceTest.cpp:
```diff
@@ -2066,9 +2076,9 @@ TEST_F(StructuralEquivalenceCacheTest, TemplateParmDepth) {
-      false, false);
+      get<0>(TU)->getASTContext().getLangOpts(), get<0>(TU)->getASTContext(),
+      get<1>(TU)->getASTContext(), NonEquivalentDecls,
+      StructuralEquivalenceKind::Default, false, false, false, false);
```
```suggestion
      StructuralEquivalenceKind::Default, /*StrictTypeSpelling=*/false, /*Complain=*/false);
```

The other two default to `false` so we can leave them out.

shafik commented on clang/unittests/AST/StructuralEquivalenceTest.cpp:
```diff
@@ -2080,9 +2090,9 @@ TEST_F(StructuralEquivalenceCacheTest, TemplateParmDepth) {
-      false, true);
+      get<0>(TU)->getASTContext().getLangOpts(), get<0>(TU)->getASTContext(),
+      get<1>(TU)->getASTContext(), NonEquivalentDecls,
+      StructuralEquivalenceKind::Default, false, false, false, true);
```
```suggestion
      StructuralEquivalenceKind::Default, /*StrictTypeSpelling=*/false, /*Complain=*/false, /*ErrorOnTagTypeMismatch=*/false, /*IgnoreTemplateParmDepth=*/true);
```

shafik commented on clang/unittests/AST/StructuralEquivalenceTest.cpp:
```diff
@@ -156,11 +158,11 @@ struct StructuralEquivalenceTest : ::testing::Test {
-        *S1.Context, *S0.Context, NonEquivalentDecls10,
-        StructuralEquivalenceKind::Default, false, false);
+        S1.Context->getLangOpts(), *S1.Context, *S0.Context,
+        NonEquivalentDecls10, StructuralEquivalenceKind::Default, false, false);
```
```suggestion
        NonEquivalentDecls10, StructuralEquivalenceKind::Default, /*StrictTypeSpelling=*/false, /*Complain=*/false);
```

shafik commented on clang/lib/Serialization/ASTReaderDecl.cpp:
```diff
@@ -4567,8 +4567,9 @@ namespace {
+              Reader.getContext().getLangOpts(), Cat->getASTContext(),
+              Existing->getASTContext(), NonEquivalentDecls,
+              StructuralEquivalenceKind::Default,
               /*StrictTypeSpelling =*/false,
               /*Complain =*/false,
               /*ErrorOnTagTypeMismatch =*/true);
```
```suggestion
              /*StrictTypeSpelling=*/false,
              /*Complain=*/false,
              /*ErrorOnTagTypeMismatch=*/true);
```

shafik reviewed (commented):
A lot of nits

shafik commented on clang/include/clang/Basic/DiagnosticASTKinds.td:
```diff
@@ -478,16 +478,24 @@ def warn_odr_function_type_inconsistent : Warning<
   "translation units (%1 vs. %2)">,
   InGroup<ODR>;
+def err_odr_attr_inconsistent : Error<
+  "attribute %0 is incompatible%select{| in different translation units}1">;
```
I can't find `is incompatible` in any of the tests.

AaronBallman replied:
Good catch, this is dead code and can be removed

shafik commented on clang/lib/AST/ASTStructuralEquivalence.cpp:
```diff
@@ -450,6 +453,116 @@ class StmtComparer {
+  // FIXME: check the attribute arguments. Attr does not track the arguments on
+  // the base class, which makes this awkward. We may want to tablegen a
+  // comparison function for attributes? In the meantime, we're doing this the
+  // cheap way by pretty printing the attributes and checking they produce
```
I am slightly uncomfortable using pretty printing, I mean theoretically we usually don't consider pretty printing load bearing as such. This is probably an ok short-term trade off but this feels potentially fragile.

cor3ntin replied:
Agreed. I also have design concerns. N3037 is silent on attributes.
Should we maybe conservatively not allow redeclarations of entities relying on compiler extensions and such?

The whole feature seems extremely brittle to maintain in the long run - and to test - so maybe an approach of limiting structural equivalence to things the standard mandates or are motivated by user request is preferable

cor3ntin replied:
consider

```c
[[deprecated]] int a;
int a; // fine 

struct S;
struct [[deprecated]] S { // fine
    int a;
};

struct S { // not fine?
    [[deprecated]] int a; // not fine?
};
```

AaronBallman replied:
> I am slightly uncomfortable using pretty printing, I mean theoretically we usually don't consider pretty printing load bearing as such. This is probably an ok short-term trade off but this feels potentially fragile.

Agreed, hence the FIXME. It's a temporary practical solution, but we don't have the underlying machinery to compare attribute arguments generically and that's far outside of the scope of this already large patch.

> Agreed. I also have design concerns. N3037 is silent on attributes.

Mismatched attributes are UB by omission, so I went with the most conservative model which is: all attributes have to be identical across objects to be structurally equivalent.

> ```
> struct S;
> struct [[deprecated]] S { // fine
>     int a;
> };
>```

Correct.

> ```
> struct S { // not fine?
>     [[deprecated]] int a; // not fine?
> };
> ```

Also correct.

uecker replied:
The standard defines when things are compatible and it explicitly does not require attributes to match for types to be compatible (in general, this has not much to do with this proposal except that this general rule now also applies here).  In general, this proposal does not raise any new questions  here (but extends the scope of such questions) and the answer to the question should be the same as elsewhere (e.g. as in typedef redefinitions, compatibility of function types). I would recommend to warn on redefinitions that add or remove standard attributes. For compatibility those should *not* matter.  For vendor attributes this up to you, of course and some attributes will certainly make types incompatible (e.g. "packed", here one can take a look at what you do for LTO because these things should already matter there).  For vendors attributes, the safe choice is to reject redefinitions. For compatibility, one has make a choice for each case, but it should be mostly obvious (does it change the representation?)

AaronBallman replied:
> The standard defines when things are compatible and it explicitly does not require attributes to match for types to be compatible.

Anything the standard doesn't talk about is UB by omission, and the standard makes no mention of attributes. There is nothing explicit about this situation.

I want to dispense with this topic because I feel very strongly we move forward with the current approach. To make this more explicit as to why:

0) The goal of this feature is that two types which were compatible across TU boundaries are now compatible if defined within the same TU. So this is about naming and object layout, in general. Attributes impact these kinds of things (packed, transparent_union, type_visibility, preferred_name, randomize_layout, etc), and attribute arguments do as well (btf_decl_tag, objc_bridge_related, uuid, etc)
1) The ideal situation is we have individual attributes specify whether they do or do not make two types [in]compatible, then the equivalence checker can generically check this rather than requiring special logic that's trivial for us to forget when adding new attributes.
2) I'm not implementing the ideal solution because that's a significant undertaking, attributes are an edge case, and I only have so much bandwidth. I have a FIXME in the patch that makes this clear.
3) Since we're not getting the ideal solution today, we can either be safe by default (require attributes to be the same) or we can be unsafe by default (ignore attributes). I'm adamant we go with safe by default until the ideal solution can be implemented.

uecker replied:
> > The standard defines when things are compatible and it explicitly does not require attributes to match for types to be compatible.
> 
> Anything the standard doesn't talk about is UB by omission, and the standard makes no mention of attributes. There is nothing explicit about this situation.

I don't think this applies here.  We say that types are compatible under conditions A B and C. That we do not mention D does not mean there is UB. We also do not mention E, F and G or a billion other things. For me, an omission is when something is not precisely specified and you can reasonable interpret it either way. This is not the case here. But this only applies to standard attributes, vendor attributes can do whatever they want.
> 
> I want to dispense with this topic because I feel very strongly we move forward with the current approach. To make this more explicit as to why:
> 
>     0. The goal of this feature is that two types which were compatible across TU boundaries are now compatible if defined within the same TU. So this is about naming and object layout, in general. Attributes impact these kinds of things (packed, transparent_union, type_visibility, preferred_name, randomize_layout, etc), and attribute arguments do as well (btf_decl_tag, objc_bridge_related, uuid, etc)
> 
>     1. The ideal situation is we have individual attributes specify whether they do or do not make two types [in]compatible, then the equivalence checker can generically check this rather than requiring special logic that's trivial for us to forget when adding new attributes.
> 
>     2. I'm not implementing the ideal solution because that's a significant undertaking, attributes are an edge case, and I only have so much bandwidth. I have a FIXME in the patch that makes this clear.
> 
>     3. Since we're not getting the ideal solution today, we can either be safe by default (require attributes to be the same) or we can be unsafe by default (ignore attributes). I'm adamant we go with safe by default until the ideal solution can be implemented.

Sounds reasonable, but I would exempt standard attributes, because I believe this is what the standard requires.

AaronBallman replied:
> I don't think this applies here.

I think it does, but I think WG14 has some serious problems around their interpretation of attributes as the feature was intended when I standardized it, at least as they relate to types. I'm not surprised we come down differently here. :-)

> Sounds reasonable, but I would exempt standard attributes, because I believe this is what the standard requires.

We make no distinction between standard and vendor attributes in our AST, they're both represented by an `Attr` subclass. And given that we have to handle things like differences between `[[noreturn]]` and `__attribute__((noreturn))`, I don't think we should exempt standard attributes.

uecker replied:
I guess then everything is UB because we never mention whether the time of day plays a role in type compatibility.

jyknight replied:
> safe by default (require attributes to be the same) or we can be unsafe by default (ignore attributes).

If we treat two types as incompatible when they ought to be compatible, it potentially could result in silent miscompilation of a valid program, since loading from memory requires that you use a type compatible with the type which was used to store. OTOH, if we treat two types as being compatible when they should be incompatible, the result is that we may incorrectly accept some invalid programs which we could've rejected. The latter is _not useful_ towards the goal of assisting the developer in writing correct code, but not unsafe on its own. 

Though...I suppose getting it wrong here may not be able to actually cause miscompilation for Clang, since we emit TBAA metadata using a different (and weaker) concept of type-compatibility anyhow.

AaronBallman replied:
> If we treat two types as incompatible when they ought to be compatible, it potentially could result in silent miscompilation of a valid program

Except this is about compatibility of types *within* the same TU, so there's no miscompile possible because the user gets an error about redefining the same type twice.

uecker replied:
> Please try to engage in the discussion in good faith rather than with snark.

Sorry, it was not meant as snark but to illustrate the point that not mentioning something explicitly in the condition does not imply that there is UB. Otherwise, what would be your rule to determine all the things that need to mentioned or not?

> The standard says (C23 6.2.7p1) "Two types are compatible types if they are the same." and then gives additional restrictions on what makes two types [in]compatible. At the end of the paragraph it says "Otherwise, the structure, union, or enumerated types are incompatible.". So the default is that two types are never compatible unless they meet a list of explicit criteria. That criteria does not mention attributes. The high-level decision tree is: are the two types the same? If not, do the two types meet these additional explicitly listed requirements? If not, they're not compatible.

The conditions which says when they are compatible does not refer to attributes, so (standard) attributes do not matter. Once the conditions is fulfill (i.e. the listed conditions are true), then those types are compatible. There is no ambiguity.

> So either this situation is UB by omission because the standard doesn't say what to do with attributes
>  or two tag types are never compatible in the presence of attributes because they're not listed as an exception to the "same type" general rule. 

If we say "A and B" are compatible when A and B share property X, then this does not imply that there are incompatible if A and B do not share another property Y. It simply means that you need to check property X while property Y is irrelevant for this question.

> UB by omission seems more reasonable to me as it gives implementations thefreedom to decide what is and isn't sufficiently compatible, which is needed anyway for vendor attributes.
> 
> (And again, the ideal goal here is to eventually let attributes handle this on an attribute-by-attribute basis.)

Yes, and for vendor attributes I think you have essentially no other choice.

AaronBallman replied:
> Therefore textual equivalence is NOT sufficient to make sure these are compatible types. Additionally, we have ''effective' attributes' that are specified by pragmas, or just inheritance that are not necessarily going to be the same thing even if textually the same.

Oh, that's an angle I hadn't thought of before: what happens when `#pragma clang attribute` slams attributes onto declarations where the attribute isn't explicitly written.

> Therefore, it is my preference/insistence that we need to white-list attributes that we know that don't alter the type. For now (as this patch/discussion has gotten large enough), it is sufficient and conservative if we just make the presence of ANY attributes on either declaration enough to make them not structurally equivalent.

I suspect this will be onerous for users in some regards given how ubiquitous attributes are, but it is defensible as an initial patch. Do you mean this to be just for declaration attributes, or type attributes as well? e.g.,
```
struct S {
  int * const bar;
  int * _Nullable Foo;
};

struct S {
  int * const bar; // Fine
  int * _Nullable Foo; // Fine, just like other qualifiers
};
```

erichkeane replied:
At the moment I'd probably want to have NO attributes including declaration attributes, but am OK white-listing certain ones (like `_Nullable`, which are always "equal").  But the same problems I mentioned above are present for both type and declaration attributes.  

THOUGH `type` attributes that modify the type then don't appear in the AST (that is, don't become an `AttributedType` and instead just create a type ala-vector-type attributes) are almost definitely fine, since we have `equivalent` defined for them already.

shafik commented on clang/lib/AST/ASTStructuralEquivalence.cpp:
```diff
@@ -450,6 +453,116 @@ class StmtComparer {
+CheckStructurallyEquivalentAttributes(StructuralEquivalenceContext &Context,
+                                      const Decl *D1, const Decl *D2,
+                                      const Decl *PrimaryDecl = nullptr) {
+  // Gather the attributes and sort them by name so that they're in equivalent
```
I am curious why you choose to ignore order. I would think we would want them to be in the same order. I find myself thinking "equivalent" should include order. Especially diagnostics are using odr in there names and I feel like that means they should be token equivalent.

AaronBallman replied:
> I am curious why you choose to ignore order.

Order within an attribute list is specified by the standard to not be important. C23 6.7.13.1p3 has this equivalent in C++: https://eel.is/c++draft/dcl.attr#grammar-4.sentence-4

shafik replied:
Ahh yes, I think I knew that at one point.

shafik commented on clang/lib/AST/ASTStructuralEquivalence.cpp:
```diff
@@ -1496,9 +1615,37 @@ static bool IsStructurallyEquivalent(StructuralEquivalenceContext &Context,
-  if (Field1->isBitField())
-    return IsStructurallyEquivalent(Context, Field1->getBitWidth(),
-                                    Field2->getBitWidth());
+  if ((Field1->isBitField() || Field2->isBitField()) &&
```
Was this a previous bug?

AaronBallman replied:
Yes, it was a previous bug and I am testing it in the n3037.c test by testing in both orders.

shafik commented on clang/lib/AST/ASTStructuralEquivalence.cpp:
```diff
@@ -1496,9 +1615,37 @@ static bool IsStructurallyEquivalent(StructuralEquivalenceContext &Context,
+          std::string Str;
+          llvm::raw_string_ostream OS(Str);
+          PrintingPolicy Policy(Context.LangOpts);
+          FD->getBitWidth()->printPretty(OS, nullptr, Policy);
```
```suggestion
          FD->getBitWidth()->printPretty(OS, /*Helper=*/nullptr, Policy);
```

shafik replied:
Also it looks like we have `getBitWidthValue()`, why not use that?

AaronBallman replied:
Good question! That interface asserts if the width is invalid whereas the `printPretty()` interface will still emit output. I went back and forth on which way to go (I wasn't able to come up with a test case where `getBitWidthValue()` would trigger an assertion, but I still wasn't comfortable going that route). I can switch if you think it's worth it.

shafik reviewed (commented):
First set of comments

shafik commented on clang/lib/AST/ASTStructuralEquivalence.cpp:
```diff
@@ -1905,54 +2093,84 @@ static bool IsStructurallyEquivalent(StructuralEquivalenceContext &Context,
+  CopyEnumerators(D1->enumerators(), D1Enums);
+  CopyEnumerators(D2->enumerators(), D2Enums);
+
+  // In C23 mode, the order of the enumerations does not matter, so sort them
```
I have to say that is a choice.

shafik commented on clang/test/C/C23/n3037.c:
```diff
@@ -0,0 +1,472 @@
+  struct T { struct S s; };  // c17-error {{redefinition of 'T'}}
+}
+
+struct food { int (*p)[3]; }; // c23-note {{field 'p' has type 'int (*)[3]' here}} \
```
Assuming `[3]` is considered the same as `[1+1+1]` in this case? Maybe add a test?

AaronBallman replied:
Sure, good call!

shafik commented on clang/test/C/C23/n3037.c:
```diff
@@ -0,0 +1,472 @@
+struct bit_field_2 { // c17-error {{redefinition of 'bit_field_2'}} \
+                        c23-error {{type 'struct bit_field_2' has incompatible definitions}}
+  int a : 1;
+  int b : 2;         // c23-note {{bit-field 'b' has bit-width 2 here}}
```
Is `2` the same as `1+1` in this case?

AaronBallman replied:
Good catch! Those are supposed to be equivalent, but they're not treated as such by the structural equivalence test. I'll fix that and add the test.

shafik commented on clang/test/C/C23/n3037.c:
```diff
@@ -0,0 +1,472 @@
+
+struct array { int y; int x[]; };    // c17-note {{previous definition is here}} \
+                                        c23-note {{field 'x' has type 'int[]' here}}
+struct array { int y; int x[0]; };   // c17-error {{redefinition of 'array'}} \
```
I just wondered, what about VLAs?

AaronBallman replied:
Can't have a VLA in a structure, so there's nothing there to test.

shafik reviewed (commented):
Second round of comments, I may make a third go around but I wouldn't wait for me.

cor3ntin commented on clang/include/clang/AST/ASTStructuralEquivalence.h:
```diff
@@ -43,6 +43,9 @@ struct StructuralEquivalenceContext {
   /// key: (from, to, IgnoreTemplateParmDepth)
   using NonEquivalentDeclSet = llvm::DenseSet<std::tuple<Decl *, Decl *, int>>;
 
+  /// The language options to use for making a structural equivalence check.
+  const LangOptions &LangOpts;
+
```
We already have 2 ASTContext; why not use the LangOpt from there? Possibly with a small function to make it less verbose.

AaronBallman replied:
Good question! I originally did that, but the problem there is: those AST contexts can be different *and can have different language options set*. So I made it explicit and force the caller to set the language options.

cor3ntin commented on clang/lib/AST/ASTStructuralEquivalence.cpp:
```diff
@@ -450,6 +453,116 @@ class StmtComparer {
+  // orders. This means that __attribute__((foo, bar)) is equivalent to
+  // __attribute__((bar, foo)).
+  llvm::SmallVector<const Attr *, 2> Attrs1, Attrs2;
+  auto CopyAttrs = [](auto &&Range, llvm::SmallVectorImpl<const Attr *> &Cont) {
+    for (const Attr *A : Range)
+      Cont.push_back(A);
+  };
```
Can you use reserve and maybe llvm::transform?

AaronBallman replied:
Yes but it's not as good as you'd hope. The issue is that the iterator for `attrs()` is a forward iterator and so getting the size means traversing the list (which means copying becomes `O(2N)`). I could use a `back_inserter` though.

cor3ntin commented on clang/lib/AST/ASTStructuralEquivalence.cpp:
```diff
@@ -450,6 +453,116 @@ class StmtComparer {
+      return II1 == II2;
+    return *II1 < *II2;
+  };
+  llvm::sort(Attrs1, Sorter);
+  llvm::sort(Attrs2, Sorter);
```
So the order doesn't matter, but the spelling does?

AaronBallman replied:
Correct, order does not matter per the standard. Spelling matters per practical considerations (`__attribute__((noreturn))` is not the same thing as `[[noreturn]]` in terms of semantics).

cor3ntin commented on clang/lib/AST/ASTStructuralEquivalence.cpp:
```diff
@@ -1454,6 +1567,12 @@ static bool IsStructurallyEquivalent(StructuralEquivalenceContext &Context,
                                      QualType Owner2Type) {
   const auto *Owner2 = cast<Decl>(Field2->getDeclContext());
 
+  // In C23 mode, check for structural equivalence of attributes on the fields.
+  // FIXME: Should this happen in C++ as well?
```
Can we get there in C++ at all? I wonder if you could assert the language mode is C
In C++ equivalence of types is limited https://eel.is/c++draft/temp.type

AaronBallman committed:
Fix comments; NFC

AaronBallman committed:
Remove unused code

AaronBallman committed:
Fix bit-width checking and add a test case

AaronBallman committed:
Use back_inserter; NFC

AaronBallman committed:
Fix formatting; NFC

jyknight commented:
It seems to me that we could simplify the implementation by being a bit more literal in the interpretation of the requirements: look only at the type, alignment, name, and bitfield-width of members. Just don't bother checking other stuff

I'd hope it can look more like a simple loop over the fields of the two record types (similar to mergeFunctionTypes's checks of the functions' param types), rather than starting with the very-complex ASTStructuralEquivalence.cpp, and adding even more complexity to validate attributes.

uecker commented:
> It seems to me that we could simplify the implementation by being a bit more literal in the interpretation of the requirements: look only at the type, alignment, name, and bitfield-width of members. Just don't bother checking other stuff
> 
> I'd hope it can look more like a simple loop over the fields of the two record types (similar to mergeFunctionTypes's checks of the functions' param types), rather than starting with the very-complex ASTStructuralEquivalence.cpp, and adding even more complexity to validate attributes.

What I do in GCC is mostly look at those things required in the standard and I make sure that all field offsets are the same (e.g. if some attribute would change how fields are layed out, this would catch it).  This is a simple loop. GCC already had a generic mechanism to compare attributes on types which decides whether an attribute makes a type incompatible (.e.g. affects ABI such as "packed"), or requires a warning, or whether it is just fine.

uecker commented:
BTW: I did not push for this feature to be exposed in earlier language modes because GCC will ship with GNU C23 by default with GCC 15.   GCC 15 already saw distribution-wide testing in Redhat and Debian and from this experience I do not expect any problems related to this feature (so far, I am not aware of any issue caused by this feature, which can not be said about some other C23 changes).  Unrelated testing exercising the new features uncovered a couple of bugs in my implementation (as well as other unrelated bugs in GCC), and I would expect that this will be true for Clang as well. But I would not expect anything critical.

AaronBallman committed:
Update test case to say this is implemented in Clang 21

AaronBallman committed:
Merge remote-tracking branch 'origin/main' into aballman-wg14-n3037

AaronBallman committed:
Add an explicit triple to the codegen test

jyknight commented:
I now see there's two different parts of the problem to worry about:

First case: A redeclaration in a _different_ scope. This always defines a new distinct type. This was valid before C23, and is still valid regardless of whether the type is compatible. (In contrast with function decls, which must be globally compatible across the entire TU in all scopes). This is the case I was thinking about earlier. This part really feels pretty-much straight-forward.

E.g. The following example defines two different `struct Foo` types. The second does not inherit any semantics from the former, and the code is valid before and after C23.
```
struct Foo { [[deprecated("hello")]] int x; };
int test2() {
  struct Foo { int x; };
  struct Foo f = {0};
  return f.x; // no deprecation
}
```

Newly in C23, the inner `struct Foo` is compatible with the outer `Foo`.  As such, you can now implicitly cast between pointers to the two types, use them to read/write the same memory, etc. The expected semantics seems to me to be very clear here: we should determine whether it's a compatible type according to the standard's rules, and we shouldn't inherit any attributes from the other scope.

For determining compatibility of nonstandard attributes, we can do similarly to how we handle functions: explicitly check a handful of properties, such as calling convention, and assume compatibility if not otherwise specified. For aggregates, we can also validate that the field offsets are the same, which will handle most of the questionable cases.

But then there's the other case...

Where you have a redefinition of a struct in the _same scope_. This was formerly invalid, and is now valid only when compatible. In this case, both definitions define "the same type", so it seems reasonable to expect some sort of decl merging semantics. But the standard doesn't appear to say anything at all about what to do.

However, we do have some prior art -- we're not inventing the whole idea of decl-merging from scratch right now. So maybe we can follow that prior art? For example, top-level attributes on a struct's definition can be inherited from a declaration in the same scope. It seems plausible that the same would be true for redefinition, and then maybe we'd also extend that same declaration-attribute-merging to fields within the struct.

Except, while investigating, I discovered that the pre-C23 behavior is inconsistent between Clang and GCC, potentially resulting in ABI differences. E.g.
```
struct __attribute__((packed)) Foo;
struct Foo {
    char a;
    int b;
};

_Static_assert(sizeof(struct Foo) == 8); // with GCC, because packed on the decl was ignored
_Static_assert(sizeof(struct Foo) == 5); // with Clang, because packed on decl was inherited by the definition
```

Both GCC and Clang do inherit deprecation from the decl. E.g.:
```
struct [[deprecated("hello")]] Foo;
struct Foo* test1; // Both GCC and Clang say: deprecated: hello

struct Foo { int x; };
struct Foo* test2; // Both GCC and Clang say: deprecated: hello
```

Given that, I'd definitely expect that under the new C23 semantics, _redefining_ the struct should obviously also preserve the deprecated attribute. Except...apparently it doesn't in GCC's implementation. Gah. Adding on to the previous test,
```
struct Foo { int x; };
struct Foo* test3; // GCC: no deprecation warning.
```

So...yeah...

uecker commented:
> I now see there's two different parts of the problem to worry about:
> 
> First case: A redeclaration in a _different_ scope. This always defines a new distinct type. This was valid before C23, and is still valid regardless of whether the type is compatible. (In contrast with function decls, which must be globally compatible across the entire TU in all scopes). This is the case I was thinking about earlier. This part really feels pretty-much straight-forward.
> 
> E.g. The following example defines two different `struct Foo` types. The second does not inherit any semantics from the former, and the code is valid before and after C23.
> 
> ```
> struct Foo { [[deprecated("hello")]] int x; };
> int test2() {
>   struct Foo { int x; };
>   struct Foo f = {0};
>   return f.x; // no deprecation
> }
> ```
> 
> Newly in C23, the inner `struct Foo` is compatible with the outer `Foo`. As such, you can now implicitly cast between pointers to the two types, use them to read/write the same memory, etc. The expected semantics seems to me to be very clear here: we should determine whether it's a compatible type according to the standard's rules, and we shouldn't inherit any attributes from the other scope.
> 
> For determining compatibility of nonstandard attributes, we can do similarly to how we handle functions: explicitly check a handful of properties, such as calling convention, and assume compatibility if not otherwise specified. For aggregates, we can also validate that the field offsets are the same, which will handle most of the questionable cases.
> 
> But then there's the other case...
> 
> Where you have a redefinition of a struct in the _same scope_. This was formerly invalid, and is now valid only when compatible. In this case, both definitions define "the same type", so it seems reasonable to expect some sort of decl merging semantics. But the standard doesn't appear to say anything at all about what to do.
> 
> However, we do have some prior art -- we're not inventing the whole idea of decl-merging from scratch right now. So maybe we can follow that prior art? For example, top-level attributes on a struct's definition can be inherited from a declaration in the same scope. It seems plausible that the same would be true for redefinition, and then maybe we'd also extend that same declaration-attribute-merging to fields within the struct.
> 
> Except, while investigating, I discovered that the pre-C23 behavior is inconsistent between Clang and GCC, potentially resulting in ABI differences. E.g.
> 
> ```
> struct __attribute__((packed)) Foo;
> struct Foo {
>     char a;
>     int b;
> };
> 
> _Static_assert(sizeof(struct Foo) == 8); // with GCC, because packed on the decl was ignored
> _Static_assert(sizeof(struct Foo) == 5); // with Clang, because packed on decl was inherited by the definition
> ```
> 
> Both GCC and Clang do inherit deprecation from the decl. E.g.:
> 
> ```
> struct [[deprecated("hello")]] Foo;
> struct Foo* test1; // Both GCC and Clang say: deprecated: hello
> 
> struct Foo { int x; };
> struct Foo* test2; // Both GCC and Clang say: deprecated: hello
> ```
> 
> Given that, I'd definitely expect that under the new C23 semantics, _redefining_ the struct should obviously also preserve the deprecated attribute. Except...apparently it doesn't in GCC's implementation. Gah. Adding on to the previous test,
> 
> ```
> struct Foo { int x; };
> struct Foo* test3; // GCC: no deprecation warning.
> ```
> 
> So...yeah...

I think this is a bug in my implementation for GCC. Here, the wording in the standard for the attributes makes this clear that an entity can be redeclared with and without the attribute, but it is considered "marked" after the first declaration that has it.

AaronBallman commented:
> However, we do have some prior art -- we're not inventing the whole idea of decl-merging from scratch right now. So maybe we can follow that prior art? For example, top-level attributes on a struct's definition can be inherited from a declaration in the same scope. It seems plausible that the same would be true for redefinition, and then maybe we'd also extend that same declaration-attribute-merging to fields within the struct.

Well, yes and no. We have prior art of doing this when we are checking valid redeclaration, but this is also about type merging as much as declaration merging. And with redefinitions, you run into problems you wouldn't be able to run into with redeclarations. e.g.,
```
struct S {
  char c;
  int x;
};

void func(struct S s) { ... } // Expects an unpacked structure

struct S {
  [[gnu::packed]] char c;
  [[gnu::packed]] int x;
};

int main() {
  struct S s = {};
  func(s); // Passes a packed structure
}
```
You couldn't run into this situation with redeclarations because you previously could not redeclare the structure fields.

The logic would sort of work out for standard attributes, at least by happenstance. There could be observable differences with `[[reproducible]]` and `[[unsequenced]]` but those are much more contrived situations given the nature of the attributes.

> So...yeah...

Yeah, there's a fair amount of moving parts. This is why I want to go with the most conservative approach of rejecting unless the attributes are the same. At some point, we'll get around to being able to tablegen more of this logic (likely leaning on the existing tablegen for some things like inheritable declaration attributes, and introducing new concepts like inheritable type attributes) and we can then relax the restrictions. But if we release in an unrestricted form, users *will* begin to rely on invalid code being accepted and we'll be in a pickle.

Curiously, GCC doesn't merge the standard attributes either, it seems to do last-one-wins: https://godbolt.org/z/j3W7ej5Kq

uecker commented:
> Curiously, GCC doesn't merge the standard attributes either, it seems to do last-one-wins: https://godbolt.org/z/j3W7ej5Kq

I filed a bug for this: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=119526

jyknight commented:
Just to say again, I think the discussion may be crisper if we consider this as two entirely-distinct features:
1. An expansion of "compatible type" for purposes of pointer casts/non-UB/etc.
2. The ability to provide two definitions of the same struct type in the same scope, to effectively behave like you can already do for a typedef.

For the former, it can be incorrect (not conservative) if we treat two types as incompatible when they should be compatible.

But, to continue discussing the latter,

> And with redefinitions, you run into problems you wouldn't be able to run into with redeclarations. e.g.,
> 
> [snip example adding packed to a field]
> 
> You couldn't run into this situation with redeclarations because you previously could not redeclare the structure fields.

You do already run into nearly that precise situation with redeclarations today, because you can (attempt to) add the packed attribute to the struct _itself_ in a redeclaration!

```
struct S {
  char c;
  int x;
};

struct [[gnu::packed]] S;
```

We handle that in `checkNewAttributesAfterDef` called from `mergeDeclAttributes`, usually emitting a warning if you try add an additional attribute on a declaration that occurs after the definition, `warning: attribute declaration must precede definition [-Wignored-attributes]`. That exact same diagnostic behavior seems like it should apply for a definition after a definition, with a little tweak of the diagnostic's wording.

Of course, that existing handling is only about top-level attributes. For fields, I think the sensible model would be to handle it as a redeclaration of the field. Then we treat the field redeclarations in the same way we handle other redeclaration situations, using the mostly-already-existing code. (We may need to add a redecl chain to FieldDecls?)

The behavior resulting from using this existing model would be: if a field declaration in a subsequent struct definition has additional attributes which the first definition did not, then we'd emit a warning (usually). If a field in a subsequent definition is _missing_ attributes a prior one had, then we'd inherit them (unless they're not inheritable). 

> Yeah, there's a fair amount of moving parts.

I think it might help to start by implementing only the expansion of "compatible type". The issues there are much simpler -- no attribute inheritance to worry about! But I don't think the PR correctly implements it yet.

Once that is settled, then go back to work on redeclaration/definition support. Possibly it'd be good to pause and ask the standards body to clarify how it's supposed to work first. While I think my proposed semantics above make sense, the standard doesn't actually _specify_ that, one way or the other. And it should.

uecker commented:
...
> > For standard attributes, the standard explicitly makes it clear that redeclarations w/ and w/o attributes are ok and can be mixed.
> 
> It does not make this clear at all. The standard is underspecified. The argument is that the standard says standard attributes are all type compatible because the standard says nothing about them while also saying that the standard says vendor attributes are UB because the standard says nothing about them. (FWIW, I continue to believe that the standard is wrong about type attributes in general and is utterly confused as to why they're used. `[[reproducible]]` and `[[unsequenced]]` do not follow the behavior of their prior art and that's a shame.)

I think it is  well specified (referenced refer to the latest draft for C2Y): It is implementation-defined what vendors attribute do (so I guess not UB): "Except where specified otherwise, the meaning of an attribute declaration is implementation-defined."  (6.7.1§10).  Redeclaration for standard attributes should be allowed by: "A name or entity declared without the nodiscard attribute can later be redeclared with the attribute and vice versa." (6.7.13.3 §3, and similar rules exist for maybe_unused and deprecated). For unsequenced and reproducible, we have in 6.7.13.8.1 "Although semantically attached to a function type, the attributes described are not part of the prototype of such a function, and redeclarations and conversions that drop such an attribute are valid and constitute compatible types...."  That standard attributes in general can not affect type compatibility follows not only from the rules for compatibility but also from "A strictly conforming program using a standard attribute remains strictly conforming in the absence of that attribute". (6.7.13.2§3) If they affect compatibility, this would not be the case anymore. We also explicitly specify what happens for composite types  "— If one of the types has a standard attribute, the composite type also has that attribute."  (6.2.7) which also only makes sense if two types with different attributes can be compatible in the first place.  It may be unfortunate that this rules are scattered across the standard and maybe there schould be a "recommended practice" for vendor attributes, but in general I do not see much room for interpretation.

uecker commented:
> > in general I do not see much room for interpretation.
> 
> The part that I find unspecified is what to do with the attributes on _fields_.
> 
> Should fields in a redefinition of a struct be treated as a redeclaration of each field from the previous definition? I _made up_ that interpretation during the course of this discussion, but I do not see it supported by any wording in the standard. Nothing you have quoted clarifies this.

What about function arguments?

I can more easily see this as something that should be mentioned specifically in the standard, but I also do not really see why one would assume different/ stricter rules for redeclarations of fields or function arguments than for other redeclarations.

AaronBallman commented:
> > > in general I do not see much room for interpretation.
> > 
> > 
> > The part that I find unspecified is what to do with the attributes on _fields_.
> > Should fields in a redefinition of a struct be treated as a redeclaration of each field from the previous definition? I _made up_ that interpretation during the course of this discussion, but I do not see it supported by any wording in the standard. Nothing you have quoted clarifies this.
> 
> What about function arguments?

It seems GCC doesn't treat function arguments the way you seem to think they should be treated: https://godbolt.org/z/E7qsnfnxo  Note how the definition of `func` calls `fp` twice while the definition of `other_func` only calls once. It seems the attribute isn't being inherited on the function parameter.

uecker commented:
> > > > in general I do not see much room for interpretation.
> > > 
> > > 
> > > The part that I find unspecified is what to do with the attributes on _fields_.
> > > Should fields in a redefinition of a struct be treated as a redeclaration of each field from the previous definition? I _made up_ that interpretation during the course of this discussion, but I do not see it supported by any wording in the standard. Nothing you have quoted clarifies this.
> > 
> > 
> > What about function arguments?
> 
> It seems GCC doesn't treat function arguments the way you seem to think they should be treated: https://godbolt.org/z/E7qsnfnxo Note how the definition of `func` calls `fp` twice while the definition of `other_func` only calls once. It seems the attribute isn't being inherited on the function parameter.

I guess it is unclear whether a redeclaration of a parameter should be considered a redeclaration of the same argument or rather a new declaration (because a new scope is opened). In any case, I do not think it matters for the question whether the redeclaration should be allowed in this case. Either it is explicitly allowed by the wording I quoted that allows redeclaration of the same name or entitty with different attributes or it is allowed simply because it is not even the same entity that is being declared. Whether GCC is consistent in using the attribute for its intended effect on optimization is a different question.

uecker commented:
> We're 83 comments into this PR and I'd like to close on the discussion so I have next steps to follow. Here's the conflicting viewpoints as I understand them:
> 
>     1. Only merge when attributes are sufficiently identical.
> 
>     2. Merge all attributes.
> 
>     3. Merge only standard attributes.
> 
>     4. Merge attributes on a case-by-case basis.
> 
>     5. Never merge when attributes are present.
> 
> 
> My position is in favor of (1) for right now and (4) eventually, @erichkeane is in favor of (5) for right now and (4) eventually, and it seems like @uecker and @jyknight are in favor of somewhere between (2) and (3) for right now and (4) eventually.
> 
> I can live with (1) or (5) for right now. (2) will lead to bugs and paints us into a corner where it will be harder to introduce constraints in the future. (4) is where I want to see us get to eventually but is way more work than I have bandwidth for.
> 
> So I think we're left with (1) which is the current patch, (3) which would require doing a fair amount of work (we don't distinguish between standard and non-standard attributes in the AST representations, so we're back to looking at lists of attributes instead of anything general), and it's work we'll be throwing away when we get to the general solution of (4), and (5) which is a simpler implementation but is very restrictive.
> 
> Given that we all want to get to (4) eventually, what's the plan for today?

I think (5) would be bad (e.g. counted_by). I have no object to clang doing (1) now as long as we get to (3) or (4) at some point later.

AaronBallman commented:
> I think (5) would be bad (e.g. counted_by). I have no object to clang doing (1) now as long as we get to (3) or (4) at some point later.

I think (5) might be too restrictive, but I fear there may be a valid reason for it as well. Same attribute with the same arguments can still mean different things, unfortunately. Consider:
```
void func(int frob, struct [[clang::objc_bridge(frob)]] S { // frob refers to the parameter here
  int x;
} s);

constexpr int frob = 12;
void func(int oh_no, struct [[clang::objc_bridge(frob)]] S { // frob refers to the global here
  int x;
} s) {

}
```
However, I had to reach pretty hard to find an example where this can happen.

erichkeane commented:
> We're 83 comments into this PR and I'd like to close on the discussion so I have next steps to follow. Here's the conflicting viewpoints as I understand them:
> 
>     1. Only merge when attributes are sufficiently identical.
> 
>     2. Merge all attributes.
> 
>     3. Merge only standard attributes.
> 
>     4. Merge attributes on a case-by-case basis.
> 
>     5. Never merge when attributes are present.
> 
> 
> My position is in favor of (1) for right now and (4) eventually, @erichkeane is in favor of (5) for right now and (4) eventually, and it seems like @uecker and @jyknight are in favor of somewhere between (2) and (3) for right now and (4) eventually.
> 
> I can live with (1) or (5) for right now. (2) will lead to bugs and paints us into a corner where it will be harder to introduce constraints in the future. (4) is where I want to see us get to eventually but is way more work than I have bandwidth for.
> 
> So I think we're left with (1) which is the current patch, (3) which would require doing a fair amount of work (we don't distinguish between standard and non-standard attributes in the AST representations, so we're back to looking at lists of attributes instead of anything general), and it's work we'll be throwing away when we get to the general solution of (4), and (5) which is a simpler implementation but is very restrictive.
> 
> Given that we all want to get to (4) eventually, what's the plan for today?

The only defensible way forward IMO is (5) today.  We can use (4) to get us to (3) (and probably should prioritize standard attributes FIRST).  

As you mentioned (2) can be really bad, but (1) has all the same problems, with a slightly different/more nuanced attack surface.  (3) is a great place to be, but is A LOT of work at least for now.  (4) is more work, but I think can be a followup.

SO I think:  (5) for now, then use (4) to get to (3) plus additional attributes we have done case-by-case.

AaronBallman committed:
Merge remote-tracking branch 'origin/main' into aballman-wg14-n3037

AaronBallman committed:
Rework attribute handling based on review feedback

If either tag type has an attribute or has a member with an attribute,
including keywords that are implemented via an attribute like _Alignas,
then the types are treated as incompatible. A warning which defaults to
an error is emitted; if the user downgrades the error back to a warning
or disables the warning entirely, then it's undefined behavior and the
user gets what they get.

Eventually we will implement tablegen logic so attributes can be
handled generically via information encoded in Attr.td.

AaronBallman committed:
Add test codegen test for attribute handling

AaronBallman commented:
I spoke with @erichkeane about this offline and ended up going with his recommended approach: any attributes cause the types to not be compatible, but instead of issuing a hard error, we issue a warning which defaults to an error. Users are free to disable the warning entirely (or downgrade it from an error) and they'll get what they get in terms of compatibility (we treat it as UB for now).

AaronBallman commented:
> I spoke with @erichkeane about this offline and ended up going with his recommended approach: any attributes cause the types to not be compatible, but instead of issuing a hard error, we issue a warning which defaults to an error. Users are free to disable the warning entirely (or downgrade it from an error) and they'll get what they get in terms of compatibility (we treat it as UB for now).

We talked about this at the Clang C and C++ Language Working Group meeting and I think the resolution we came to was:

* There are two scenarios we need to care about: same type in same scope, looking to see whether the second definition is the same as the first definition; different type in different scopes, looking to see whether the two types are compatible.
* For the first scenario, we want to look at attributes and reject if any are present. Eventually this will be relaxed and handled via tablegen code on a case-by-case basis.
* For the second scenario, we don't need to look at attributes explicitly, we could use the record layout as a stand-in for compatibility (do the structural check ignoring attributes, if they're structurally equivalent, do a layout check to ensure things like alignment match up). However, this too eventually needs to be tightened and handled via tablegen code on a case-by-case basis.

So the changes needed to move this patch forward are to only perform the attribute check when we're checking for a valid redefinition, not type compatibility in general. A follow-up to improve upon this can do the layout check when looking for compatibility. Alternatively (or as another follow-up), we can implement the case-by-case basis logic.

Does that match your understanding @jyknight?

AaronBallman committed:
Merge remote-tracking branch 'origin/main' into aballman-wg14-n3037

AaronBallman committed:
Add another test case

these changes.

AaronBallman committed:
Oops, does not actually fix this issue.

The diagnostic has always appeared on Windows and never on Linux, and
that hasn't changed with this patch.

AaronBallman committed:
Merge remote-tracking branch 'origin/main' into aballman-wg14-n3037

AaronBallman commented:
Ping

AaronBallman commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -11399,6 +11400,22 @@ static QualType mergeEnumWithInteger(ASTContext &Context, const EnumType *ET,
   return {};
 }
 
+QualType ASTContext::mergeTagTypes(QualType LHS, QualType RHS) {
```
I'm following the naming convention already used by other helpers: `mergeExceptionSpecs`. `mergeFunctionTypes`, `mergeFunctionParameterTypes`, and so on. The intent is that this will eventually do actual merging logic (so the name is reasonable), but even for right now, you either get the "merged" type (which is either `LHS` or `RHS`) or you get `QualType{}`. WDYT?

erichkeane replied:
What about `mergeTagDefinitions`?  I don't have a good idea here otherwise, but it isn't really merging tag-types so much as just their definitions?

Also perhaps even in NOT the C2(whatever) mode: what about returning the 'not definition' (assuming 1 is a definition and 1 isnt?).  it makes it a no-op in that case, but perhaps one that is sensible.

AaronBallman commented on clang/lib/AST/ASTStructuralEquivalence.cpp:
```diff
@@ -450,6 +453,41 @@ class StmtComparer {
+  // the same semantic attribute, differences in attribute arguments, order
+  // in which attributes are applied, how to merge attributes if the types are
+  // structurally equivalent, etc.
+  const Attr *D1Attr = nullptr, *D2Attr = nullptr;
```
Yeah, I think we should just hold off until we're ready to bite the bullet and do it properly.

AaronBallman committed:
mergeTagTypes -> mergeTagDefinitions

AaronBallman committed:
Merge remote-tracking branch 'origin/main' into aballman-wg14-n3037

AaronBallman commented:
> So the changes needed to move this patch forward are to only perform the attribute check when we're checking for a valid redefinition, not type compatibility in general. A follow-up to improve upon this can do the layout check when looking for compatibility. Alternatively (or as another follow-up), we can implement the case-by-case basis logic.
> 
> Does that match your understanding @jyknight?

Ping on this @jyknight

uecker commented:
I did some initial testing by compiling one of my projects and so far this worked fine.

AaronBallman commented:
> > So the changes needed to move this patch forward are to only perform the attribute check when we're checking for a valid redefinition, not type compatibility in general. A follow-up to improve upon this can do the layout check when looking for compatibility. Alternatively (or as another follow-up), we can implement the case-by-case basis logic.
> > Does that match your understanding @jyknight?
> 
> Ping on this @jyknight

Another ping (if no response by mid-week, I'll probably land the changes as-is and address concerns post-commit).

AaronBallman committed:
Merge remote-tracking branch 'origin/main' into aballman-wg14-n3037

uecker commented:
So one usability issue I noticed that we also had in GCC is that the warning about declaring structs in function prototypes (-Wvisibility) does not make sense anymore and becomes annoying. We turned it off in C23 for complete (but not for incomplete) structs) without tag.

AaronBallman commented:
> So one usability issue I noticed that we also had in GCC is that the warning about declaring structs in function prototypes (-Wvisibility) does not make sense anymore and becomes annoying. We turned it off in C23 for complete (but not for incomplete) structs) without tag.

https://github.com/llvm/llvm-project/pull/138516

cyndyishida commented on clang/lib/AST/ASTStructuralEquivalence.cpp:
```diff
@@ -450,6 +453,41 @@ class StmtComparer {
+  if (D2->hasAttrs())
+    D2Attr = *D2->getAttrs().begin();
+  if (D1Attr || D2Attr) {
+    const auto *DiagnoseDecl = cast<TypeDecl>(PrimaryDecl ? PrimaryDecl : D2);
```
For some more context, we've been qualifying the 21.x branch and discovered that the series of patches related to implementing WG14 N3037, has broken several Objective-C projects with `gnu23` set. Building with clang modules is a common configuration that seems to have been broken here by these structural equivalence checks. 

These are the relevant commits we found that represent the regression. It is basically the set of commits I needed to be able to revert cleanly. 
```
460ff1eb0ef10a599b96a5a7fa1f4bd3973de4c7 - [C23] AST equivalence of attributes (#151196)
3d6fb12dfcef05407439580491430c376489d6a5 - [C23] More improved type compatibility for enumerations (#150946)
3b53c84e338bfb870f49e6a520725d86ff40c375 - C23] Handle type compatibility for enumerations better (#150282)
6769a836e97d5a09b239f78a8fd63cf4dac1fe13 -  [C23] Handle type compatibility of unnamed records (#141783)
ee3610efb511ec0a8d5ebd2016ce23b4b4344cd1  -  Remove unused function; NFC
a4ceac7e3e04c32cb3e334eb89b54d8e99a298f8  - [C23] Implement WG14 N3037 (#132939)
```

Akira has a patch to simply account for `Context.Complain` when diagnosing violations, but that further exposes an issue with modular builds. I've attached a reproducer that contains a set of files & a modulemap to demonstrate the issue. `run.sh` just invokes clang; you can change the path to point to what is correct on your system and assumes you are in the `repro` directory. 

[repro.zip](https://github.com/user-attachments/files/22646854/repro.zip)

The reproduced case basically is that when we are compiling a TU (client.m) that itself is a part of the same project that vends a module. The compiler accepts an option `-fmodule-name=Foo` to let the compiler know this. When this happens, headers from that module are treated textually, but if those same headers are resolved transitively through a different module, that would be respected as a modular header & thus `Foo.pcm` is created and also contains the same declarations. This typically would result in a redeclaration issue, but [the compiler explicitly supports this case through type merging.](https://github.com/llvm/llvm-project/blob/e84dcba9246d696715fe1daa4ddb218182580a70/clang/lib/Serialization/ASTReaderDecl.cpp#L1362) It seems like for the unnamed struct that is in ivar, the structural equivalence check now fails, but previously didn't, due to https://github.com/llvm/llvm-project/blob/e84dcba9246d696715fe1daa4ddb218182580a70/clang/lib/AST/ASTStructuralEquivalence.cpp#L1774 checking `!D1->getIdentifier()` which is the case for unnamed structs. 

We do heavily rely on type merging support & this configuration is working. @AaronBallman could you please take a look? Apologies in advance for the delay in reporting the observed breakages.
